src/main/java/org/datadog/jenkins/plugins/datadog/traces/BuildSpanManager.java

```diff
@@ -1,34 +1,50 @@
 package org.datadog.jenkins.plugins.datadog.traces;
 
-import org.datadog.jenkins.plugins.datadog.traces.message.TraceSpan;
-
-import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.logging.Logger;
+import org.datadog.jenkins.plugins.datadog.traces.message.TraceSpan;
 
 /**
- * Used to propagate the build Span between onStart() and onComplete() methods.
- * This mechanism is needed because the Span object cannot be serialized in a Jenkins Action.
+ * Used to store trace data after the build has finished.
+ * The data is needed to link upstream build to a downstream build.
  */
 public class BuildSpanManager {
 
+    private static final Logger LOGGER = Logger.getLogger(BuildSpanManager.class.getName());
+
     private static final BuildSpanManager INSTANCE = new BuildSpanManager();
-    private final Map<String, TraceSpan> traceSpanByBuildTag = new HashMap<>();
+    private final Map<String, TraceSpan.TraceSpanContext> contextByTag = new ConcurrentHashMap<>();
+    private final BlockingQueue<String> tags = new ArrayBlockingQueue<>(getCapacity());
```

Comment:
`ArrayBlockingQueue` is thread-safe, right?

Reply:
Yes, all `BlockingQueue` impls are thread-safe